dm/dm/worker/server.go

```diff
@@ -94,129 +98,152 @@ func NewServer(cfg *Config) *Server {
 func (s *Server) Start() error {
 	log.L().Info("starting dm-worker server")
 	RegistryMetrics()
-	s.ctx, s.cancel = context.WithCancel(context.Background())
-	tls, err := toolutils.NewTLS(s.cfg.SSLCA, s.cfg.SSLCert, s.cfg.SSLKey, s.cfg.AdvertiseAddr, s.cfg.CertAllowedCN)
-	if err != nil {
-		return terror.ErrWorkerTLSConfigNotValid.Delegate(err)
-	}
-
-	rootLis, err := net.Listen("tcp", s.cfg.WorkerAddr)
-	if err != nil {
-		return terror.ErrWorkerStartService.Delegate(err)
-	}
-	s.rootLis = tls.WrapListener(rootLis)
 
-	s.etcdClient, err = clientv3.New(clientv3.Config{
-		Endpoints:            GetJoinURLs(s.cfg.Join),
-		DialTimeout:          dialTimeout,
-		DialKeepAliveTime:    keepaliveTime,
-		DialKeepAliveTimeout: keepaliveTimeout,
-		TLS:                  tls.TLSConfig(),
-		AutoSyncInterval:     syncMasterEndpointsTime,
-	})
-	if err != nil {
-		return err
-	}
+	var m cmux.CMux
 
-	s.setWorker(nil, true)
+	// protect member from data race. some functions below like GetRelayConfig,
+	// GetSourceBoundConfig has a built-in timeout so it will not be stuck for a
+	// long time.
+	startErr := func() error {
+		if ok := s.ctxMu.Lock(s.ctx); !ok {
```

Comment:
it makes code more complex, do we really need this ctxmutex? s.ctx is cancel only on close

Reply:
it's about other s.Lock should be aware of s.ctx done. please check PR discription

Reply:
I remeber the stack shows 

https://github.com/pingcap/tiflow/blob/a1da9b42ec10284c49d276ae932f6f858a02ad92/dm/dm/worker/server.go#L162-L172

s..observeSourceBound is blocked on s.Lock, so it can't s.wg.Done()

Reply:
we need 1 mutex to sync between operations and 1(may not need) to sync accessing fields. and another mutex to sync start/close. but now there's only 1 mutex.

still trying to accept the idea of `context mutex`, odd to me.

Reply:
Maybe using multiple mutex and a good locking order we also can avoid deadlock as well. Will take a look soon

Reply:
https://github.com/pingcap/tiflow/pull/6213/commits/66cc7422e11240a5ba15df5902d9a9c57280f678